File: app/src/main/java/com/chess/cryptobot/model/response/bittrex/BittrexResponse.java
package com.chess.cryptobot.model.response.bittrex;

import com.chess.cryptobot.model.Price;
import com.chess.cryptobot.model.response.AddressResponse;
import com.chess.cryptobot.model.response.BalanceResponse;
import com.chess.cryptobot.model.response.CurrenciesListResponse;
import com.chess.cryptobot.model.response.CurrenciesResponse;
import com.chess.cryptobot.model.response.MarketResponse;
import com.chess.cryptobot.model.response.OrderBookResponse;
import com.chess.cryptobot.model.response.PaymentResponse;
import com.chess.cryptobot.model.response.TickerResponse;
import com.chess.cryptobot.model.response.TradeLimitResponse;
import com.chess.cryptobot.model.response.TradeResponse;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class BittrexResponse implements MarketResponse,
        BalanceResponse,
        OrderBookResponse,
        CurrenciesListResponse,
        AddressResponse,
        PaymentResponse,
        TradeResponse,
        TradeLimitResponse
{

    private Boolean success;
    private String message;
    private final BittrexGenericResponse[] results;

    BittrexResponse(BittrexGenericResponse[] results) {
        this.results = results;
    }

    BittrexResponse(BittrexGenericResponse result) {
        this.results = new BittrexGenericResponse[1];
        results[0] = result;
    }

    void setSuccess(Boolean success) {
        this.success = success;
    }

    @Override
    public boolean success() {
        return success;
    }

    public void setMessage(String message) {
        this.message = message;
    }

    @Override
    public String message() {
        return message;
    }

    @Override
    public Double getAmount() {
        return results[0].getAvailable();
    }

    @Override
    public List<Price> bids() {
        return parsePrices(results[0].getBuy());
    }

    @Override
    public List<Price> asks() {
        return parsePrices(results[0].getSell());
    }

    public List<TickerResponse> getTickers() {
        return Arrays.asList(results);
    }

    private List<Price> parsePrices(List<BittrexPrice> prices) {
        ArrayList<Price> parsedPrices = new ArrayList<>();
        prices.forEach(price -> parsedPrices.add(new Price(price.getRate(), price.getQuantity())));
        return parsedPrices;
    }

    @Override
    public List<CurrenciesResponse> getInfo() {
        return Arrays.asList(results);
    }

    @Override
    public String getAddress() {
        return results[0].getAddress();
    }

    @Override
    public String getPaymentId() {
        return results[0].getUuid();
    }

    @Override
    public String getTradeId() {
        return results[0].getUuid();
    }

    @Override
    public Double getTradeLimitByName(String pairName) {
        for(BittrexGenericResponse result: results) {
            if (result.getPairName().equals(pairName)) {
                return result.getMinTradeSize();
            }
        }
        return null;
    }
}

File: app/src/main/java/com/chess/cryptobot/model/response/bittrex/BittrexResponse.kt
package com.chess.cryptobot.model.response.bittrex

import com.chess.cryptobot.model.History
import com.chess.cryptobot.model.Price
import com.chess.cryptobot.model.response.*
import java.util.function.Consumer

class BittrexResponse : MarketResponse, BalanceResponse, OrderBookResponse, CurrenciesListResponse, AddressResponse, TradeLimitResponse {
    private var success: Boolean? = null
    private var message: String? = null
    val results: Array<BittrexGenericResponse?>

    internal constructor(results: Array<BittrexGenericResponse?>) {
        this.results = results
    }

    internal constructor(result: BittrexGenericResponse?) {
        results = arrayOfNulls(1)
        results[0] = result
    }

    fun setSuccess(success: Boolean?) {
        this.success = success
    }

    override fun success(): Boolean {
        return success!!
    }

    fun setMessage(message: String?) {
        this.message = message
    }

    override fun message(): String? {
        return message
    }

    override val amount: Double
        get() = results[0]!!.available

    override fun bids(): List<Price?>? {
        return parsePrices(results[0]!!.buy)
    }

    override fun asks(): List<Price?>? {
        return parsePrices(results[0]!!.sell)
    }

    val tickers: List<TickerResponse>
        get() = results.toList().filterIsInstance<TickerResponse>()

    private fun parsePrices(prices: List<BittrexPrice>?): List<Price?> {
        val parsedPrices = ArrayList<Price?>()
        prices!!.forEach(Consumer { price: BittrexPrice -> parsedPrices.add(Price(price.rate!!, price.quantity!!)) })
        return parsedPrices
    }

    override fun getCurrencies(): List<CurrenciesResponse> {
       return results.toList().filterIsInstance<CurrenciesResponse>()
    }

    override val address: String?
        get() = results[0]!!.address

    override fun getTradeLimitByName(pairName: String?): Double? {
        for (result in results) {
            if (result!!.pairName == pairName) {
                return result.minTradeSize
            }
        }
        return null
    }

    val history: List<History>
        get() = HistoryResponseFactory(results.toList().filterIsInstance<HistoryResponse>()).history

}